### ISSUE TITLE
Allow other encodings

### ISSUE THREAD

**Issue by JordenVerwer** [whatwg/encoding#207]
Right now, the specification states "User agents must not support any other encodings or labels". This seems overly restrictive to me. While I understand there could be reasons for not allowing specific encodings for security reasons, many unsupported encodings (IBM437, IBM850 and IBM858, to name just a few) are completely harmless. The gap between "must support" (IBM866) and "must not support" (IBM437) is rather huge.

I would suggest one of
1. not forbidding user agents from supporting other encodings, or
2. adding a list of encodings that must not be supported (with encodings that are not on either list being optionally supported), or
3. adding a list of optionally supported encodings.

I'm specifically not asking you to add more mandatory encodings (although I wouldn't object to it in case of IBM437 either). If you're going to add a list of optionally supported encodings, please include IBM437 in it, which I consider to be a highly valuable encoding for web browsers to support, especially for plain text files, but also for HTML files (because they do get encoded as IBM437 on rare occasions).

The reason I'm asking this is that right now, the wording of the specification is a reason for browser makers explicitly not to add other encodings, including IBM437.

**Comment by annevk** [comment]
It's not harmless of the set of legacy encodings that browsers support is different between browsers. That causes interoperability issues. That's where we came from, we're not going back there.

**Comment by JordenVerwer** [comment]
Going along with your line of reasoning for the sake of argument, I have to make a couple of remarks. I'll alternate between prescriptive and descriptive points of view. First of all, I can understand the desire to make browsers behave more uniformly, but in this particular case I think you may be taking things one step too far. Not every aspect can be 100% identical, not every aspect is 100% identical and not every aspect will be 100% identical. In case of IBM437, some browsers support it (Mobile: Edge; Desktop: Chromium-based browsers on Linux, Safari, the old Edge and Internet Explorer) whereas some don't (Mobile: all browsers I checked on iOS and Android; Desktop: Chromium-based browsers on Windows (including the new Edge) and Firefox). Secondly, while the spec does say that other encodings should not be supported by user agents, it does not (to the best of my knowledge) say how this non-support should work out in practice. The purest solution would be to display an error message instead of the requested page, but that may be too draconic. At the very least I would expect a warning message saying the page may not be displayed correctly, but none of the browsers I checked exhibit this behavior. Finally, the fact that there is a list of supported legacy encodings at all must be some kind of compromise. I could understand (but not fully agree with) a decision to support UTF-8 only, but instead there is a pretty long list of encodings that frankly, seems (and objectively speaking probably is) rather arbitrary. So even if I were to accept your premise that different browsers supporting different encodings is somehow harmful (which I don't), I'm not seeing that idea having been worked out in a consistent manner in the current specification.

Because I get the feeling that I'm not going to convince you to go with any of my three suggestions, alternatively I'd like to ask you to add IBM437 to the list of supported encodings. This is because this encoding is very prevalent among text files originating from the early eighties to about the mid-nineties and sees some usage in HTML documents as well. Many users, including me, also use their browsers to view text files they find online and I would rather not suggest that browsers support IBM437 for text files but not for HTML files. Besides, the ability to view IBM437-encoded HTML files is useful in itself.

I expect my claim that IBM437 used to be a common (if not the most common) encoding for text files to be uncontroversial, but I realize that some might state that its use in HTML files is very rare. I do encounter IBM437-encoded HTML files once in a while and tried to look for objective research to prove my point. Unfortunately I did not find it, although I didn't find any refutation either. The closest I came to objective numbers relates to usenet and email:

http://quetzalcoatal.blogspot.com/2014/03/understanding-email-charsets.html

I'm not advocating adding any other encodings along with IBM437, but (hypothetically) if you were to say "If we include IBM437 we have to include IBM850 too, or else the spec will become inconsistent" or something along those lines, I would not oppose to it either. I only want to make it clear that I'm not advocating adding a sundry list of encodings to the spec, just IBM437.

**Comment by annevk** [comment]
How "page loading" works is defined by HTML and it does go into your questions to some extent, though not as far as I would like. The Encoding Standard primarily concerns itself with mapping labels to encodings and byte sequences to scalar values (and vice versa).

Looking at https://en.wikipedia.org/wiki/Code_page_437#Character_set there's a good reason to never support that: https://encoding.spec.whatwg.org/#security-background.

**Comment by hsivonen** [comment]
> some browsers support it 
...
> Desktop: Chromium-based browsers on Linux

@inexorabletash, what is this about? Does Chromium allow building with system ICU such that there's no Chromium-side allow-list?

> This is because this encoding is very prevalent among text files originating from the early eighties to about the mid-nineties and sees some usage in HTML documents as well. Many users, including me, also use their browsers to view text files they find online

Do these text files declare IBM437? If not, how would viewing English text files that use the non-ASCII parts of IBM437 only for box drawing work in browsers that don't have a character encoding menu? (Presumably, an encoding detector would want to err on the side of the box drawing bytes being non-ASCII letters from windows-125x, since those cases are for more common on the Web.)

I understand that there is certain historical appeal to e.g. textfiles.com serving the original bytes instead of converting to UTF-8 on the server side. However, all files that I've encountered there that use the non-ASCII parts of IBM437 use parts that are the same in IBM866, so they'd work in browsers if declared as IBM866.

> Looking at https://en.wikipedia.org/wiki/Code_page_437#Character_set there's a good reason to never support that: https://encoding.spec.whatwg.org/#security-background.

FWIW, we already replaced the low half of IBM866 with ASCII. (I'm not advocating for extending the set of supported encodings. If we ever change the set of supported encodings, I'd rather drop IBM866 than add more DOS encodings.)

**Comment by annevk** [comment]
Oh interesting, now I wonder if we should add a Notes column to the single-byte encodings table for interesting tidbits like that.

**Comment by hsivonen** [comment]
> http://quetzalcoatal.blogspot.com/2014/03/understanding-email-charsets.html

This was written by a Thunderbird developer. Since then, Thunderbird has moved to supporting fewer encodings: Only those in the Encoding Standard plus UTF-7 and, in addition to the Encoding Standard label set, supporting Java 1.1 labels and a small number of other labels for these encodings.

In practice, Thunderbird didn't get complaints about IBM437. It did get a bug files about there existing at least one IBM Notes installation that was configured to send ISO-2022-KR, but Thunderbird didn't restore ISO-2022-KR support.

**Comment by hsivonen** [comment]
> Oh interesting, now I wonder if we should add a Notes column to the single-byte encodings table for interesting tidbits like that.

If we add notes, let's also include the kind of stuff that the docs for encoding_rs has for each encoding. [Example for EUC-KR](https://docs.rs/encoding_rs/0.8.22/encoding_rs/static.EUC_KR.html).

**Comment by hsivonen** [comment]
(Specifically, the bit about code page 949 having PUA mappings. Though I haven't seen any practical consequences, it bothers me EUC-KR in the spec is different from what Windows does on that point.)

**Comment by JordenVerwer** [comment]
> How "page loading" works is defined by HTML and it does go into your questions to some extent, though not as far as I would like. The Encoding Standard primarily concerns itself with mapping labels to encodings and byte sequences to scalar values (and vice versa).
> Looking at https://en.wikipedia.org/wiki/Code_page_437#Character_set there's a good reason to never support that: https://encoding.spec.whatwg.org/#security-background.

Assuming you are talking about the control character glyphs, I agree that it would be wise to explicitly define this: I support ignoring the glyphs and interpreting all code points under 128 as ASCII.

> Do these text files declare IBM437? If not, how would viewing English text files that use the non-ASCII parts of IBM437 only for box drawing work in browsers that don't have a character encoding menu? (Presumably, an encoding detector would want to err on the side of the box drawing bytes being non-ASCII letters from windows-125x, since those cases are for more common on the Web.)

Generally, an HTTP server will include headers indicating the use of IBM437 if it is configured correctly. Browsers that respect those headers will show the box drawing characters correctly.

> I understand that there is certain historical appeal to e.g. textfiles.com serving the original bytes instead of converting to UTF-8 on the server side. However, all files that I've encountered there that use the non-ASCII parts of IBM437 use parts that are the same in IBM866, so they'd work in browsers if declared as IBM866.

Well, I suppose you could pretend a file is encoded as IBM866 if it is encoded as IBM437 but does not use any of the code points in which the two differ, but that seems rather hackish to me. I can also imagine the use of diacritics and/or Greek letters in IBM437-encoded texts along with the box drawing characters, although I don't have an immediate example in front of me. I also note that there are multiple variations of IBM866, so without looking into the details of the version you selected I can't say for certain if it contains all mathematical symbols included in IBM437.

**Comment by inexorabletash** [comment]
> @inexorabletash, what is this about? Does Chromium allow building with system ICU such that there's no Chromium-side allow-list?

There's a USING_SYSTEM_ICU build flag that's been around forever. No Chrome builds use this, of course.

**Comment by hsivonen** [comment]
> I can also imagine the use of diacritics and/or Greek letters in IBM437-encoded texts along with the box drawing characters, although I don't have an immediate example in front of me.

That's only a step away from asking for all the DOS encodings to be added instead of just IBM437. That binary size impact adds up with very little practical benefit.

> There's a USING_SYSTEM_ICU build flag that's been around forever. No Chrome builds use this, of course.

It seems bad for that option not to enforce an allow-list on the Chromium side.

**Comment by JordenVerwer** [comment]
> That's only a step away from asking for all the DOS encodings to be added instead of just IBM437. That binary size impact adds up with very little practical benefit.

I don't think it is quite that bad, but I do understand your concerns. That's one reason I'm not asking you to add them all. Right now a single IBM* encoding is supported, so supporting a subset of them is not unprecedented. I'm only asking for the addition of a second one, so the scope of my request is very limited. Across all text files taken together (not just HTML), I'm pretty sure IBM437 is the most prevalent DOS-era encoding.

I'm not afraid of a slippery slope. Realistically speaking, someone might come along and ask for IBM850 and maybe even IBM00858 (which is just IBM850 with the dotless i replaced by the Euro sign), but I can't imagine anyone seriously requesting the addition of any more of them.

So in the very worst case, you'd end up supporting four IBM* encodings, but honestly I think it would stop at just IBM866 and IBM437.

**Comment by domenic** [comment]
I think if anything we should be trying to drive down the number of legacy encodings supported and remove them, not increase the number.

**Comment by JordenVerwer** [comment]
What would make the addition of IBM437 acceptable to you?

**Comment by hsivonen** [comment]
> What would make the addition of IBM437 acceptable to you?

Likely nothing.

As far as I can tell, Firefox has [never supported IBM437](https://bugzilla.mozilla.org/show_bug.cgi?id=298436). Later other DOS encodings in the IBM8xy range were added for the OS/2 port. I suspect they were more about dealing with the system than with Web content.

Firefox dropped the IBM8xy encodings other than IBM866 in Firefox 16 (patch landed on 2012-11-10, release on 2013-02-19). By then, Chrome already didn't support those encodings.

Since then, there's been one [Thunderbird report](https://bugzilla.mozilla.org/show_bug.cgi?id=1587813) about IBM437 and [two](https://bugzilla.mozilla.org/show_bug.cgi?id=846936) [Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1296306) reports about IBM850.

While I understand that you care about this particular encoding, evidence suggests that as a matter of the big picture, support for legacy Web content doesn't (in serious quantity) require IBM437 to be supporte

**Comment by JordenVerwer** [comment]
As an individual, I guess my request to add support for an actual encoding that was introduced with a solid technical rationale at the time and that has been used in massive quantities, even though that was mostly before the rise of the web, will not be fulfilled. Similar requests from others are admittedly rare, but not nonexistent. Some of the examples you pointed out were actually new to me, so thank you for that.

So, against my better judgment, I'll contrast that to a hypothetical scenario: if I had been in the government of a large country, I could just have made up my own encoding over a decade after the invention of UTF-8 with absolutely no technical justification, then mandated its use, in which case you would likely have added it to the list of "legacy" encodings that must be supported by all browsers. Of course, any similarity between my hypothetical scenario and actual events is entirely coincidental... ;)

In short, although we disagree, I do understand where you're coming from. I just find it hard to look at a laundry list of encodings that must be supported, many of which have no legitimate raison d'être at all in my opinion, while IBM437 will not be. If UTF-8 had been the sole supported encoding I would likely just have accepted defeat, but now I simply can't shake the feeling that the decision was arbitrary.

**Comment by hsivonen** [comment]
> In short, although we disagree, I do understand where you're coming from. I just find it hard to look at a laundry list of encodings that must be supported, many of which have no legitimate raison d'être at all in my opinion, while IBM437 will not be. If UTF-8 had been the sole supported encoding I would likely just have accepted defeat, but now I simply can't shake the feeling that the decision was arbitrary.

Browsers used to add encodings without strong use cases. For example, ISO-8859-15 was always about political posturing and not about solving a problem that windows-1252 wasn't already solving. In general, some encodings is the ISO-8859 series came along just by being part of the series without being evaluated individually. 

The Encoding Standard looks arbitrary, because it tried to freeze the intersection of what major browsers supported: i.e. it's an intersection of semi-arbitrary decisions.

In retrospect, by the time the Encoding Standard was written, we might have gotten away with dropping IBM866, ISO-8859-3, ISO-8859-10, ISO-8859-14, and ISO-8859-16, but actually investigating whether we could drop these is not worth the trouble. This still isn't a reason to add more.
